keps/sig-storage/0037-20190129-PVCTaint.md

```diff
@@ -0,0 +1,123 @@
+
+Previously, we proposed to use Readiness Gate for this problem, however, this approach lacks the support the scenarios that some special Pods (e.g., Pods for populating the data) can be allowed to use the volume even when the data is not ready.
+
+Taint and toleration feature currently used for marking node and reacting based
+on the taint effect seems also provides a nice way to solve this volume issue.
+
+### Goals
+
+This proposal is trying to investigate the taint and toleration feature applied
+to PVC/PV. It will identify the use cases that this feature could be applied to
+and describe the workflow of how this feature can be used.
+
+### Non-Goals
+
+
+## Proposal
+
+Currently, Kubernetes already supports three types of taint effects: 
+- NoSchedule: Pods that do not tolerate this taint are not scheduled on the node.
+- PreferNoSchedule: Kubernetes avoids scheduling Pods that do not tolerate this taint onto the node.
+- NoExecute: Pod is evicted from the node if it is already running on the node, and is not scheduled
+onto the node if it is not yet running on the node.
+
+These effects are used for tainting the nodes so that scheduler or controller
+will take actions on Pods based on the effect. Here we propose a new effect, TaintEffectNoAttachOrMount. For a Pod that cannot tolerate this taint, attach_detach controller and volume manager should skip this volume and not trigger attach or mount operations.
+
+For example,
+```
+taint := &v1.Taint{
+	Key:    dataNotReady,
+	Value:  "true",
+	Effect: v1.TaintEffectNoAttachOrMount,
+}
+```
+
+We have choices of either tainting the PVC or PV. 
+
+###Taint PV
+
+- For dynamic provisioning: when provisioner creates an empty volume, it can check if the data has not populated from the specified data source into the volume and taint the PV with the NoAttachorMount taint. Because of this taint, when attach_detach_controller or volume manager receives this volume, it will not trigger volume operation unless the Pod which references this volume has a toleration of the taint. 
+The data populator controller should run in a Pod with a toleration of the taint so that it can attach and mount the volume and then populate the data to the volume. After the data is populated, the controller can remove the DataNotReady taint from the PV.
+
+- For static provisioning, system admin could taint the PV when applicable. Similar to dynamic provisioning, after data is populated, the taint will be removed from the PV by the controller.
+
+
+### Taint PVC
+
+It might be also beneficial to allow users to taint a PVC for other purposes, e.g., to avoid the PVC is used immediately by some pod. In this case, the attach_detach_controller or volume manager should check the taint for PVC besides PV and take actions accordingly.  In addition, data populator controller needs to remove the taint from either PVC or PV
+
+
+## User Stories
```

Comment:
I am thinking that a PV taint could also be potentially used for volume health monitoring.  If a driver reports a volume as unhealthy, then that could potentially trigger some response to evict the Pod using it.

But it's currently just a thought, nothing really concrete at this point.  I mention it here as a potential future use case we should consider when designing the API.